File: zig/atbash-cipher/atbash_cipher.zig
const std = @import("std");
const mem = std.mem;

/// Encodes `s` using the Atbash cipher. Caller owns the returned memory.
pub fn encode(allocator: mem.Allocator, s: []const u8) mem.Allocator.Error![]u8 {
    var ans = std.ArrayList(u8).init(allocator);

    var i: u8 = 0;
    for (s) |c| {
        const new_c = switch (c) {
            'a'...'z' => 'z' - (c - 'a'),
            'A'...'Z' => 'z' - (c - 'A'),
            '0'...'9' => c,
            else => continue,
        };
        if (i == 5) {
            try ans.append(' ');
            i = 0;
        }
        try ans.append(new_c);

        i += 1;
    }
    return ans.toOwnedSlice();
}

/// Decodes `s` using the Atbash cipher. Caller owns the returned memory.
pub fn decode(allocator: mem.Allocator, s: []const u8) mem.Allocator.Error![]u8 {
    var ans = std.ArrayList(u8).init(allocator);
    for (s) |c| {
        const new_c = switch (c) {
            'a'...'z' => 'z' - (c - 'a'),
            '0'...'9' => c,
            else => continue,
        };
        try ans.append(new_c);
    }
    return ans.toOwnedSlice();
}

File: cpp/atbash-cipher/atbash_cipher.cpp
#include "atbash_cipher.h"

#include <cctype>

namespace atbash_cipher {
string encode(string s) {
    string ans = "";
    for (size_t i = 0; i < s.length(); i++) {
        char ch = s.at(i);
        if (isalpha(ch)) {
            ans += 'z' - tolower(ch) + 'a';
        } else if (isdigit(ch)) {
            ans += ch;
        }

        if (ans.length() % 6 == 5) {
            ans += ' ';
        }
    }

    if (ans.length() > 1 && ans[ans.length() - 1] == ' ') {
        ans.pop_back();
    }
    return ans;
}

string decode(string s) {
    string ans = "";
    for (auto ch : s) {
        if (isalpha(ch)) {
            ans += 'z' - tolower(ch) + 'a';
        } else if (isdigit(ch)) {
            ans += ch;
        }
    }
    return ans;
}
}  // namespace atbash_cipher
